Add loading and running recipes to the notebook API

stefsmeets opened the pull request:
Hi everyone, this PR implements an api to load and run recipes in an interactive jupyter notebook.

This PR is mostly complete in terms of what it is supposed to do. The thing that is missing is to have good test. I'm thinking we should have some very basic test recipes that we can run to test the API.

Also implements bugfix that closes #953

**To do**
- [x] Look at issue with multi-processing (`max_parallel_tasks != 1`) -> Seems to work as intended.
- [x] Implement unit tests with basic recipe for testing (`.yml` or maybe pass dict directly?)
- [x] https://github.com/ESMValGroup/ESMValTool_sample_data/pull/15 merged
- [x] https://github.com/ESMValGroup/ESMValTool_sample_data/pull/16 merged and release/depend on v0.0.3

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Follows up on #901 and #868 
For more discussion, see issue #498 
Documentation: https://esmvaltool--907.org.readthedocs.build/projects/ESMValCore/en/907/api/esmvalcore.api.html 
Demo notebook: https://gist.github.com/stefsmeets/1163eeea765906550f772c15b69e866e

stefsmeets commented:
Hi @nielsdrost and @Peter9192 , this PR is ready for review. Could you have a look?

There is still a small issue with the tests on CircleCI that I will have a look at, but I think other than that the PR can be reviewed. The tests are running fine locally.

stefsmeets commented:
Tests are fixed, just one unrelated to this PR 👍

```
self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>

    def test_regrid__cell_specification(self):
        specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
        scheme = 'linear'
        for spec in specs:
            result = regrid(self.src_cube, spec, scheme)
            self.assertEqual(result, self.regridded_cube)
>           self._check(spec, scheme, spec=True)

tests/unit/preprocessor/_regrid/test_regrid.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
    self.assertEqual(_CACHE[spec], self.tgt_grid)
E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140715029656144'>
```

Peter9192 commented on esmvalcore/experimental/utils.py:
```diff
@@ -64,21 +66,36 @@ def get_all_recipes(subdir: str = None) -> list:
         for filename in filenames:
-            try_path = Path(location, filename)
+            try_path = Path(location, filename).expanduser()
             if try_path.exists():
```
Perhaps it would be nice to print here something like "Found recipe_blablalba.yml" in "interesting/path". In that way, we could avoid any confusion as to which recipe is being used if there happen to be two of them when only 1 is expected or vice versa.

stefsmeets replied:
Good idea, I added the search locations to the `FileNotFoundError`.

Peter9192 reviewed (commented):
Hey Stef!
Very nice work so far!! I had a first look at the code - didn't test anything yet. But I have a couple of suggestions and questions. My main concern is that I'm a bit worried that the distinction between RecipeInfo and Recipe is very soon going to be very confusing. I'd rather make them be the same thing, i.e. the API is just a wrapper around the internal Recipe object. I left some ideas to that effect, maybe we can discuss it later.

stefsmeets commented on esmvalcore/config-user.yml:
```diff
@@ -37,6 +37,25 @@ config_developer_file: null
 # Get profiling information for diagnostics
 # Only available for Python diagnostics
 profile_diagnostic: false
+# Maximum number of years to use.
```
These are set in main by the CLI, but default values must be defined somewhere, and having them in the default user config is the most logical place.

https://github.com/ESMValGroup/ESMValCore/blob/9b6f0955d56e18f4e989435f25fe05b12d773f88/esmvalcore/_main.py#L302

stefsmeets commented on doc/api/esmvalcore.api.recipe_info.rst:
```diff
@@ -12,9 +12,54 @@ Recipe metadata
+    >>> output = recipe_info.run(session)
+    <log messages>
+
+:py:meth:`esmvalcore.experimental.RecipeInfo.run` returns an object that contains the locations of the data and figures (not implemented yet).
```
Yeah, at the moment it returns nothing. Either way works for me, but it is also a reminder that the functionality is not complete yet.

stefsmeets commented on esmvalcore/experimental/_logging.py:
```diff
@@ -0,0 +1,48 @@
+
+
+@contextmanager
+def log_to_dir(drc: str) -> None:
```
`log_to_dir` was the best I could come up with. I updated the docstring to make it more clear what this does.

stefsmeets commented on tests/sample_data/experimental/recipe_api_test.yml:
```diff
@@ -0,0 +1,28 @@
+      ta:
+        preprocessor: preproc
+        mip: Amon
+    scripts:
```
This is already covered by the rest of the test suite.

My goal here is to have the simplest test I could come up with (suggestions welcome!) to cover the new functionality. But it does enable more complete testingusing real recipes, which is something we may want in the future.

stefsmeets commented on setup.py:
```diff
@@ -60,7 +60,7 @@
         'pytest-xdist',
         ('ESMValTool_sample_data @ '
-         'git+https://github.com/ESMValGroup/ESMValTool_sample_data@v0.0.1'),
+         'git+https://github.com/ESMValGroup/ESMValTool_sample_data@367929e'),
```
Yeah, this was temporary while debugging the CircleCI issue.

stefsmeets commented on esmvalcore/experimental/config/_config_object.py:
```diff
@@ -163,6 +164,19 @@ def config_dir(self):
         """Return user config directory."""
         return USER_CONFIG_DIR
 
+    def to_config_user(self) -> dict:
```
The difference is that esmvalcore takes a dictionary. It calls this dictionary `config_user` (I'm not super happy with this name), which is essentially a copy of the global config (defined here: https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/_config.py) with the session directories added.

The same object is called `Session` in the api, but it is not a dictionary. Therefore a session must be converted to a dict for compatilibity with recipes. I considered just calling the method `.to_dict` at first, but I think `.to_config_user` makes the intent more clear.

There are two options I think:
- Use session objects everywhere (this needs master to depend on the new config, we tried this before)
- Make a compatibility layer (which is what is done in this PR)

stefsmeets commented on esmvalcore/experimental/recipe_info.py:
```diff
@@ -57,8 +75,13 @@ def from_tag(cls, tag: str):
 
 class Project:
-    """Contains project information."""
+    """Contains project information.
```
I specifically did not want this to depend on `config-references`, so it does not take a `TAG` directly. Tags can be resolved using the `from_tag` method.

Peter9192 replied:
```suggestion
    """Contains acknowledgement information related to this recipe.
```

stefsmeets commented on esmvalcore/experimental/recipe_info.py:
```diff
@@ -150,12 +189,13 @@ def render(self, renderer: str = 'plaintext') -> str:
 class RecipeInfo():
```
We will need to use `esmvalcore.Recipe` anyway to get the diagnostics / scripts, so I like the idea of inheriting from it or delegating to it. It would need a bit of refactoring to get this to work then.

 I think from a user point of view, it only makes sense to call it `Recipe` and call the `esmvalcore.Recipe` something like `RecipeDriver` or `RecipeEngine`.

Peter9192 replied:
Let's make a new issue to refactor the esmvalcore.Recipe class, so that it will be easier to inherit from it in this API module.

stefsmeets replied:
#934

Peter9192 replied:
Didn't we agree to call this object a plain old 'Recipe' here?

stefsmeets commented on esmvalcore/experimental/recipe_info.py:
```diff
@@ -273,3 +318,60 @@ def references(self) -> tuple:
+                        config_user=config_user,
+                        recipe_file=self.path)
+
+        return recipe
```
Yeah, I think this is a better solution.

stefsmeets commented:
> Very nice work so far!! I had a first look at the code - didn't test anything yet. But I have a couple of suggestions and questions. My main concern is that I'm a bit worried that the distinction between RecipeInfo and Recipe is very soon going to be very confusing. I'd rather make them be the same thing, i.e. the API is just a wrapper around the internal Recipe object. I left some ideas to that effect, maybe we can discuss it later.

Hi @Peter9192 , I went through your suggestions just now. Not all of them are straightforward to answer at this point. I agree with you that `RecipeInfo` will most likely turn into `Recipe` very soon with the planned work. I have some ideas on how to go about it, but perhaps  this is something to look at together for a follow-up PR.

stefsmeets commented:
Hi @Peter9192 , have I addressed your comments? If so, could you please approve this PR?

From my point of view, this is ready to merge. But I would like to see if there is any on #936 before we do so.

stefsmeets commented:
Hi @bouweandela , could you have a look at the changes to `config-user.yml` and let me know if they are in line with #936? 

And do you have any idea what is the cause of this error? It seems to be completely unrelated to the changes in the code.

```
self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>

    def test_regrid__cell_specification(self):
        specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
        scheme = 'linear'
        for spec in specs:
            result = regrid(self.src_cube, spec, scheme)
            self.assertEqual(result, self.regridded_cube)
>           self._check(spec, scheme, spec=True)

tests/unit/preprocessor/_regrid/test_regrid.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
    self.assertEqual(_CACHE[spec], self.tgt_grid)
E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
```

Peter9192 commented:
Hi @stefsmeets I thought we decided to make a PR for #934 and use it here, but maybe I'm confusing things now? I'd say the sooner we do that, the better. How do you see this now?

stefsmeets commented:
> 
> 
> Hi @stefsmeets I thought we decided to make a PR for #934 and use it here, but maybe I'm confusing things now? I'd say the sooner we do that, the better. How do you see this now?

It is more necessary for the follow-up item, to get some output back. I don't see any advantages for this PR.

Peter9192 reviewed (changes requested):
I tested it now and it seems to work well. I'm fine with doing #934 after this one if you feel more comfortable with that, I just thought the earlier we do it the better, since we're still flexible.

If you could just look once more at the thing that I unresolved about the naming of the RecipeInfo class? Otherwise I'm happy with the implementation.

Peter9192 commented:
Oh and by the way the tests pass locally for me as well

stefsmeets commented:
> I tested it now and it seems to work well. I'm fine with doing #934 after this one if you feel more comfortable with that, I just thought the earlier we do it the better, since we're still flexible.
> 
> If you could just look once more at the thing that I unresolved about the naming of the RecipeInfo class? Otherwise I'm happy with the implementation.

I think I got everything, should be good now 👍 can you press approve if you are happy? 😁

Peter9192 commented:
I guess we cannot merge until codacy passes though...

stefsmeets commented:
> Hi @bouweandela , could you have a look at the changes to `config-user.yml` and let me know if they are in line with #936?

After our discussion I will not define these in `config-user.yml`, I will find another location. Maybe `config_defaults.py`?

> And do you have any idea what is the cause of this error? It seems to be completely unrelated to the changes in the code.
> 
> ```
> self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>
> 
>     def test_regrid__cell_specification(self):
>         specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
>         scheme = 'linear'
>         for spec in specs:
>             result = regrid(self.src_cube, spec, scheme)
>             self.assertEqual(result, self.regridded_cube)
> >           self._check(spec, scheme, spec=True)
> 
> tests/unit/preprocessor/_regrid/test_regrid.py:107: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
>     self.assertEqual(_CACHE[spec], self.tgt_grid)
> E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
> ```

stefsmeets commented:
This test fail seems to be completely random on CircleCI (sometime it passes, sometimes it fails), unrelated to any changed made in this PR, and I cannot reproduce it on my own machine.

> ```
> self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>
> 
>     def test_regrid__cell_specification(self):
>         specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
>         scheme = 'linear'
>         for spec in specs:
>             result = regrid(self.src_cube, spec, scheme)
>             self.assertEqual(result, self.regridded_cube)
> >           self._check(spec, scheme, spec=True)
> 
> tests/unit/preprocessor/_regrid/test_regrid.py:107: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
>     self.assertEqual(_CACHE[spec], self.tgt_grid)
> E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
> ```

bouweandela commented on esmvalcore/_task.py:
```diff
@@ -21,6 +21,23 @@
+    return dumper.represent_scalar('tag:yaml.org,2002:str', str(data))
+
+
+def check_levels_representer(dumper, data):
+    """For CheckLevel enums in yaml files."""
+    return dumper.represent_scalar('tag:yaml.org,2002:str', str(data))
+
+
+yaml.representer.SafeRepresenter.add_representer(Path, path_representer)
+yaml.representer.SafeRepresenter.add_representer(PosixPath, path_representer)
+yaml.representer.SafeRepresenter.add_representer(CheckLevels,
+                                                 check_levels_representer)
```
```suggestion
yaml.representer.SafeRepresenter.add_representer(Path, path_representer)
yaml.representer.SafeRepresenter.add_representer(PosixPath, path_representer)
```
Why do you need to represent `CheckLevels` as yaml?

stefsmeets replied:
Good catch, this is probably no longer needed if I do not need to have the config parameters in the `config-user.yml` 👍

bouweandela commented on esmvalcore/_recipe.py:
```diff
@@ -302,16 +303,10 @@ def _get_default_settings(variable, config_user, derive=False):
-        'frequency': variable['frequency'],
-        'check_level': config_user['check_level']
-    }
+    settings['cmor_check_data'] = settings['cmor_check_metadata']
```
```suggestion
    settings['cmor_check_data'] = dict(settings['cmor_check_metadata'])
```
I would feel safer if these were not the same dict, though it will probably work fine as is

stefsmeets replied:
Yeah good point, I was wondering about this too!

bouweandela reviewed (changes requested):
Code looks good, though I did not try using it.

Could you have a look at the question about representing CheckLevels in a YAML file? I think this should not be needed, because CheckLevels should not leak to the diagnostic script through the YAML interface, this is supposed to be used by the CMOR checker and nothing else.

Would it be possible to solve the import outside toplevel Codacy issue?

stefsmeets commented:
> Would it be possible to solve the import outside toplevel Codacy issue?

Yeah, it can be solved, but I think it is clearer where it is now.

bouweandela commented:
> Yeah, it can be solved, but I think it is clearer where it is now.

If would be great if you could follow the style guide if there is no technical reason for deviating from it. If everyone starts making up their own style the code will become harder to read for everyone.

stefsmeets commented:
> 
> 
> > Yeah, it can be solved, but I think it is clearer where it is now.
> 
> If would be great if you could follow the style guide if there is no technical reason for deviating from it. If everyone starts making up their own style the code will become harder to read for everyone.

I don't consider this error be a 'hard' fail from Codacy, sometimes the code reads much better if the import is not at the top of the file. Besides that, Codacy and yapf sometimes conflict as well, so it is unclear what style to use.

stefsmeets commented:
This PR is ready to be merged. I worked around the error by clearing the cache before the regrid tests, but a better implementation is needed. For more info see #953.

stefsmeets commented:
Shoot, now I merged the wrong branch here...
